xds/src/main/java/com/linecorp/armeria/xds/ClusterResourceNode.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.xds;
+
+import static com.linecorp.armeria.xds.XdsType.CLUSTER;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Objects;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.envoyproxy.envoy.config.cluster.v3.Cluster;
+import io.envoyproxy.envoy.config.core.v3.ConfigSource;
+import io.envoyproxy.envoy.config.endpoint.v3.ClusterLoadAssignment;
+import io.envoyproxy.envoy.config.route.v3.Route;
+import io.envoyproxy.envoy.config.route.v3.VirtualHost;
+
+final class ClusterResourceNode extends AbstractResourceNode<ClusterSnapshot>
+        implements SnapshotWatcher<EndpointSnapshot> {
+
+    @Nullable
+    private final VirtualHost virtualHost;
+    @Nullable
+    private final Route route;
+    private final int index;
+
+    ClusterResourceNode(@Nullable ConfigSource configSource,
+                        String resourceName, XdsBootstrapImpl xdsBootstrap,
+                        @Nullable ResourceHolder primer, SnapshotWatcher<? super ClusterSnapshot> parentNode,
```

Comment:
> How about creating a SnapshotWatcher<EndpointSnapshot> as a private static class for simplicity?

In order to make it a `static` class since the watcher needs to be aware of the most up-to-date snapshot. I feel like using a holder object for the `ResourceHolder` is overkill.

I've pushed 4f0bfd7ebff732cff9f96f8ac1c8d4fe4cea8ebd which uses inner classes at the cost of more object allocation, let me know if this looks like what you expected